File: javascript/easy/Middle_of_the_Linked_List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const middleNode = function (head) {
  let middle = head;
  let end = head;

  while (end != null && end.next != null) {
    middle = middle.next;
    end = end.next.next;
  }

  return middle;
};

File: java/easy/Middle_of_the_Linked_List.java
package easy;

public class Middle_of_the_Linked_List {


	 // Definition for singly-linked list.
	 public class ListNode {
	      int val;
	      ListNode next;
	      ListNode() {}
	      ListNode(int val) { this.val = val; }
	      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
	  }
	 

	class Solution {
	    public ListNode middleNode(ListNode head) {
	        ListNode middleNode = head;
	        ListNode endNode = head;

	        while(endNode != null && endNode.next != null) {
	            middleNode = middleNode.next;
	            endNode = endNode.next.next;
	        }

	        return middleNode;
	    }
	}
}
